### PR TITLE
Arbitrary LocalDateTimes

### PR DESCRIPTION
**Pull request by gtrefs**
First draft for #1776 

This pull requests adds a new generator for `LocalDateTime`s which is used by `Arbitrary`.

### CODE REVIEW

File: javaslang-test/src/main/java/javaslang/test/Arbitrary.java [hunk 1]
```diff
@@ -197,6 +199,18 @@
...
+    }
+
+    static Arbitrary<LocalDateTime> localDateTime(ChronoUnit unit) {
+        return localDateTime(LocalDateTime.now(), unit);
+    }
+
+    static Arbitrary<LocalDateTime> localDateTime(LocalDateTime median, ChronoUnit unit) {
+        return size -> random -> Gen.choose(median.minus(size, unit), median.plus(size, unit)).apply(random);
+    }
+
```
Comment (danieldietrich) [inline javaslang-test/src/main/java/javaslang/test/Arbitrary.java]: My first thought was why we are not returning the generator directly? I think the min/max bounds need to be calculated only once. I see that I did the same for Arbitrary.integer(). But I think it does not make sense...

```java
return size -> Gen.choose(median.minus(size, unit), median.plus(size, unit));
```

File: javaslang-test/src/main/java/javaslang/test/Gen.java [hunk 2]
```diff
@@ -222,6 +225,19 @@
...
+        if (end.isBefore(start)) {
+            throw new IllegalArgumentException("End date is before start date");
+        }
+        return random -> {
+            final long duration = Duration.between(start, end).toMillis();
+            final long from = Gen.choose(0, duration).apply(random);
+            return start.plus(from, ChronoUnit.MILLIS);
+        };
+    }
+
```
Comment (danieldietrich) [inline javaslang-test/src/main/java/javaslang/test/Gen.java]: I think Gen is the wrong place for this method. Gen is a general construct for specifying how to generate random values of primitive types and generic objects.

More complex logic can (and should) build solely on this.

See my next comment how to move this to Arbitrary...

File: javaslang-test/src/main/java/javaslang/test/Arbitrary.java [hunk 3]
```diff
@@ -197,6 +199,18 @@
...
+    static Arbitrary<LocalDateTime> localDateTime() {
+        return localDateTime(ChronoUnit.DAYS);
+    }
+
+    static Arbitrary<LocalDateTime> localDateTime(ChronoUnit unit) {
+        return localDateTime(LocalDateTime.now(), unit);
+    }
+
+    static Arbitrary<LocalDateTime> localDateTime(LocalDateTime median, ChronoUnit unit) {
+        return size -> random -> Gen.choose(median.minus(size, unit), median.plus(size, unit)).apply(random);
```
Comment (danieldietrich) [inline javaslang-test/src/main/java/javaslang/test/Arbitrary.java]: We move the logic from Gen to Arbitrary like this:

```java
Objects.requireNonNull(median, "median is null");
Objects.requireNonNull(unit, "unit is null");
return size -> {
    final LocalDateTime start = median.minus(size, unit);
    final LocalDateTime end = median.plus(size, unit);
    final long duration = Duration.between(start, end).toMillis();
    final Gen<Long> from = Gen.choose(0, duration);
    return random -> start.plus(from.apply(random), ChronoUnit.MILLIS);
}
```

_(Note: We name the parameters directly in the error messages)_

Some Gen.localDateTime tests might be obsolete or need to be moved to ArbitraryTest...

Comment (gtrefs) [inline javaslang-test/src/main/java/javaslang/test/Arbitrary.java]: Thanks for the comments. I will add the changes later :)

Comment (danieldietrich) [inline javaslang-test/src/main/java/javaslang/test/Arbitrary.java]: great!

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java [hunk 4]
```diff
@@ -185,6 +189,79 @@ public int nextInt(int bound) {
...
+    public void shouldCreateArbitraryLocalDateTime(){
+        final Arbitrary<LocalDateTime> date = Arbitrary.localDateTime();
+
+        assertThat(date).isNotNull();
+    }
+
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotAcceptNullMedianLocalDateTime(){
+        final Arbitrary<LocalDateTime> dates = Arbitrary.localDateTime(null, ChronoUnit.DAYS);
```
Comment (sputnikci) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: \[FindBugs\] INFO: DLS: Dead store to dates in javaslang\.test\.ArbitraryTest\.shouldNotAcceptNullMedianLocalDateTime\(\)

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java [hunk 5]
```diff
@@ -185,6 +189,79 @@ public int nextInt(int bound) {
...
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotAcceptNullChronoUnit(){
+        final Arbitrary<LocalDateTime> dates = Arbitrary.localDateTime(LocalDateTime.now(), null);
```
Comment (sputnikci) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: \[FindBugs\] INFO: DLS: Dead store to dates in javaslang\.test\.ArbitraryTest\.shouldNotAcceptNullChronoUnit\(\)

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java [hunk 6]
```diff
@@ -185,6 +189,79 @@ public int nextInt(int bound) {
...
+
+        Property.def("With size of 100 days, dates should be in range of +/- 100 days")
+                .forAll(arbitrary)
+                .suchThat(d -> d.isAfter(median.minusDays(100)) && d.isBefore(median.plusDays(100)))
+                .check(-100, 1000);
+    }
+
+    @Test
+    public void shouldGenerateTwoDifferentSuccessiveDates(){
+        final LocalDateTime end = LocalDateTime.now().minusDays(2);
```
Comment (sputnikci) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: \[FindBugs\] INFO: DLS: Dead store to end in javaslang\.test\.ArbitraryTest\.shouldGenerateTwoDifferentSuccessiveDates\(\)

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java [hunk 7]
```diff
@@ -185,6 +189,79 @@ public int nextInt(int bound) {
...
+        final LocalDateTime start = LocalDateTime.now().minusDays(3);
```
Comment (sputnikci) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: \[FindBugs\] INFO: DLS: Dead store to start in javaslang\.test\.ArbitraryTest\.shouldGenerateTwoDifferentSuccessiveDates\(\)

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java (hunk 4 above)
Comment (danieldietrich) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: When you test exceptional cases (like here) please perform just a call - the variable is not needed, i.e.

```java
@Test(expected = NullPointerException.class)
public void shouldNotAcceptNullMedianLocalDateTime(){
    Arbitrary.localDateTime(null, ChronoUnit.DAYS);
}
```

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java (hunk 5 above)
Comment (danieldietrich) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: dito

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java (hunk 6 above)
Comment (danieldietrich) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: this line can be removed (variable `end` not used)

File: javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java (hunk 7 above)
Comment (danieldietrich) [inline javaslang-test/src/test/java/javaslang/test/ArbitraryTest.java]: this line can be removed (variable `start` not used)

**Comment by gtrefs** [comment]
It was a bit late yesterday and I had some craft beers :smile:. Sorry for not cleaning before pushing. I hope this is fine now.

**Comment by danieldietrich** [comment]
> I had some craft beers 😄

beer only is ok, and always remember: wine after beer is fine ;-p

Looks great to me! Waiting now for the CI server, then will pull it in.

**Comment by gtrefs** [comment]
Don't push it yet. I accidently removed some empty lines in `Arbitrary`

**Comment by danieldietrich** [comment]
Ok, will wait!

**Comment by danieldietrich** [comment]
Looks awesome, thank you!

**Pull request merged**